starboard/shared/starboard/player/job_thread.h

```diff
@@ -83,13 +82,44 @@ class JobThread {
 
     return job_queue_->RemoveJobByToken(job_token);
   }
-
  private:
+  ~JobThread();
   static void* ThreadEntryPoint(void* context);
   void RunLoop();
 
   pthread_t thread_;
   std::unique_ptr<JobQueue> job_queue_;
+
+  friend class ScopedJobThreadPtr;
+};
+
+// The ScopedJobThreadPtr class guarantees that the pointer to JobThread object
+// is valid during JobThread destructor. This prevents issues of accessing nullified
+// JobThread pointer, as per b/372515171
+class ScopedJobThreadPtr {
```

Comment:
Can you explain exactly how this class differs in behavior from a `std::unique_ptr<JobThread>`? I don't see a difference. If I am missing a subtle difference, please tell me exactly where.

Reply:
std::unuque_ptr<> on destructor first nullifies the pointer it holds, and then call the delete/destructor of said pointer. This is the problem specifically for JobThread, since the thread code might attempt to access that pointer before the ~JobThread() is done.

Reply:
I think that means that the PR lets the thread access an object that is being destructed, which is worse than dereferencing a nullptr. 

It seems that the bad pattern there is the work done in the destructor. What if that includes calling a virtual function implemented in a derived class that already finished its destructor when this one runs? I'm guessing in our case here we're getting callbacks from the thread, triggered from the destructor, into methods here that check if there is a thread. But technically, it's in the process of being destructed so it already may partially be destroyed.

I can see how the PR may avoid the issue being seen, but it just serves to hide the real issue. The work  that triggers the callbacks shouldn't be done from the destructor.

Reply:
I can see it been problematic in general case, but this class works only with this implementation of JobThread.